File: contract-java-test/src/test/java/examples/client/SecretSharingDownload.java
package examples.client;

import static java.nio.charset.StandardCharsets.UTF_8;

import com.partisiablockchain.BlockchainAddress;
import java.math.BigInteger;

/**
 * Minimal DEMO program for download secret-sharings to secret-sharing contract.
 *
 * <p>Uses a predefined secret-key, and blockchain reader node.
 *
 * <p>Argument format: {@code <SHARING MODE> <CONTRACT ADDRESS> <ID: NUM>}
 *
 * <p>Where {@code <SHARING MODE>} is either {@code "xor"} or {@code "shamir"}
 */
public final class SecretSharingDownload {

  /**
   * Main method.
   *
   * @param args CLI arguments. Not nullable.
   */
  public static void main(String[] args) {
    // Load arguments
    final String sharingMode = args[0];
    final SecretShares.Factory<?> factory = SecretSharingUpload.getFactory(sharingMode);
    final BlockchainAddress contractAddress = BlockchainAddress.fromString(args[1]);
    final BigInteger shareId = new BigInteger(args[2]);

    // Download secret sharing
    final byte[] reconstructedSecret =
        SecretSharingUpload.secretSharingClient(contractAddress, factory)
            .downloadAndReconstruct(shareId);

    // Print it
    System.out.println(new String(reconstructedSecret, UTF_8));
  }
}

File: ts/src/test/examples/client/SecretSharingDownload.ts
import { BlockchainAddress } from "@partisiablockchain/blockchain-api-transaction-client";
import BN from "bn.js";
import { SecretShareFactory } from "../secretsharing/SecretShareFactory";
import { SecretShares } from "../secretsharing/SecretShares";
import { getFactory, getSharingClient } from "./SecretSharingUpload";

/**
 * Minimal DEMO program for download secret-sharings to secret-sharing contract.
 *
 * <p>Uses a predefined secret-key, and blockchain reader node.
 *
 * <p>Argument format: <code><SHARING MODE> <CONTRACT ADDRESS> <ID: NUM></code>
 *
 * <p>Where <code><SHARING MODE></code> is either <code>xor</code> or <code>shamir</code>
 */
async function main() {
  if (process.argv.length !== 5) {
    console.error(
      `Invalid number of arguments. Expected '3' but received ${process.argv.length - 2}. Correct format is:\n` +
        "npm run download-shares <SHARING MODE> <CONTRACT ADDRESS> <SHARE ID>"
    );
    return;
  }
  const sharingMode = process.argv[2];
  const factory: SecretShareFactory<SecretShares> = getFactory(sharingMode);
  const contractAddress: BlockchainAddress = process.argv[3];
  const id: BN = new BN(process.argv[4]);

  const sharingClient = getSharingClient(contractAddress, factory);
  const secret = await sharingClient.downloadAndReconstruct(id);
  // eslint-disable-next-line no-console
  console.log(secret.toString("utf-8"));
}

if (require.main === module) {
  main();
}
